File: src/api/collections.ts
/**
 * Collections API
 */

import { invoke } from '@tauri-apps/api/core';
import type { Collection } from '../types/collection';

/**
 * Create a new collection
 */
export async function createCollection(
  name: string,
  collectionType: 'manual' | 'smart',
  filterQuery?: string,
  icon?: string,
  color?: string
): Promise<Collection> {
  return await invoke<Collection>('create_collection', {
    name,
    collectionType,
    filterQuery,
    icon,
    color,
  });
}

/**
 * Get all collections
 */
export async function getCollections(): Promise<Collection[]> {
  return await invoke<Collection[]>('get_collections');
}

/**
 * Get a specific collection
 */
export async function getCollection(id: string): Promise<Collection> {
  return await invoke<Collection>('get_collection', { id });
}

/**
 * Update a collection
 */
export async function updateCollection(
  id: string,
  name?: string,
  icon?: string,
  color?: string,
  filterQuery?: string
): Promise<Collection> {
  return await invoke<Collection>('update_collection', {
    id,
    name,
    icon,
    color,
    filterQuery,
  });
}

/**
 * Delete a collection
 */
export async function deleteCollection(id: string): Promise<void> {
  return await invoke('delete_collection', { id });
}

/**
 * Add a document to a collection
 */
export async function addDocumentToCollection(
  documentId: string,
  collectionId: string
): Promise<void> {
  return await invoke('add_document_to_collection', {
    documentId,
    collectionId,
  });
}

/**
 * Remove a document from a collection
 */
export async function removeDocumentFromCollection(
  documentId: string,
  collectionId: string
): Promise<void> {
  return await invoke('remove_document_from_collection', {
    documentId,
    collectionId,
  });
}

/**
 * Get all documents in a collection
 */
export async function getCollectionDocuments(collectionId: string): Promise<string[]> {
  return await invoke<string[]>('get_collection_documents', { collectionId });
}

/**
 * Get all collections for a document
 */
export async function getDocumentCollections(documentId: string): Promise<Collection[]> {
  return await invoke<Collection[]>('get_document_collections', { documentId });
}

/**
 * Preview smart collection results
 */
export async function getSmartCollectionPreview(filterQuery: string): Promise<string[]> {
  return await invoke<string[]>('get_smart_collection_preview', { filterQuery });
}

/**
 * Reorder collections
 */
export async function reorderCollections(collectionIds: string[]): Promise<void> {
  return await invoke('reorder_collections', { collectionIds });
}

File: src-tauri/src/commands/collections.rs
//! Collections commands for organizing documents

use crate::models::collection::{Collection, CollectionType, DocumentCollection, SmartCollectionFilter};
use std::sync::Arc;

#[tauri::command]
pub async fn create_collection(
    name: String,
    collection_type: String,
    filter_query: Option<String>,
    icon: Option<String>,
    color: Option<String>,
) -> Result<Collection, String> {
    let collection_type_enum = CollectionType::from_str(&collection_type)
        .ok_or_else(|| format!("Invalid collection type: {}", collection_type))?;

    let mut collection = Collection::new(name);
    collection.collection_type = collection_type_enum;
    collection.filter_query = filter_query;
    collection.icon = icon;
    collection.color = color;

    // TODO: Save to database
    Ok(collection)
}

#[tauri::command]
pub async fn get_collections() -> Result<Vec<Collection>, String> {
    // TODO: Fetch from database
    // Return default collections for now
    Ok(vec![
        Collection::to_read(),
        Collection::in_progress(),
        Collection::completed(),
        Collection::favorites(),
        Collection::recent(),
    ])
}

#[tauri::command]
pub async fn get_collection(id: String) -> Result<Collection, String> {
    // TODO: Fetch from database
    Err("Not implemented".to_string())
}

#[tauri::command]
pub async fn update_collection(
    id: String,
    name: Option<String>,
    icon: Option<String>,
    color: Option<String>,
    filter_query: Option<String>,
) -> Result<Collection, String> {
    // TODO: Update in database
    Err("Not implemented".to_string())
}

#[tauri::command]
pub async fn delete_collection(id: String) -> Result<(), String> {
    // TODO: Delete from database
    Err("Not implemented".to_string())
}

#[tauri::command]
pub async fn add_document_to_collection(
    document_id: String,
    collection_id: String,
) -> Result<(), String> {
    // TODO: Add document_collection record to database
    Err("Not implemented".to_string())
}

#[tauri::command]
pub async fn remove_document_from_collection(
    document_id: String,
    collection_id: String,
) -> Result<(), String> {
    // TODO: Remove document_collection record from database
    Err("Not implemented".to_string())
}

#[tauri::command]
pub async fn get_collection_documents(collection_id: String) -> Result<Vec<String>, String> {
    // TODO: Fetch document IDs for this collection
    // For smart collections, run the filter query
    Ok(vec![])
}

#[tauri::command]
pub async fn get_document_collections(document_id: String) -> Result<Vec<Collection>, String> {
    // TODO: Fetch all collections containing this document
    Ok(vec![])
}

#[tauri::command]
pub async fn get_smart_collection_preview(
    filter_query: String,
) -> Result<Vec<String>, String> {
    // TODO: Run the filter query and return matching document IDs
    Ok(vec![])
}

#[tauri::command]
pub async fn reorder_collections(
    collection_ids: Vec<String>,
) -> Result<(), String> {
    // TODO: Update sort order for collections
    Err("Not implemented".to_string())
}
